lib/ecto/adapters/sql.ex

```diff
@@ -434,72 +511,160 @@ defmodule Ecto.Adapters.SQL do
     {pool, timeout} = pool!(repo)
 
     opts    = Keyword.put_new(opts, :timeout, timeout)
-    timeout = Keyword.get(opts, :timeout)
-    worker  = checkout_worker(pool, timeout)
+    timeout = Keyword.fetch!(opts, :timeout)
+    key     = {:ecto_info, pool}
+
+    case Process.get(key) do
+      nil ->
+        begin_transaction(repo, pool, key, fun, opts, timeout)
+      info ->
+        transaction(repo, key, info, fun, opts, timeout)
+    end
+  end
 
+  defp begin_transaction(repo, pool, key, fun, opts, timeout) do
+    info = ask(pool, timeout)
+    _ = Process.put(key, info)
     try do
-      do_begin(repo, worker, opts)
-      value = fun.()
-      do_commit(repo, worker, opts)
-      {:ok, value}
+      transaction(repo, key, info, fun, opts, timeout)
+    after
+      done(Process.delete(key))
+    end
+  end
+
+  defp transaction(repo, key, info, fun, opts, timeout) do
+    try do
+      do_begin(repo, key, info, opts, timeout)
+      fun.()
     catch
       :throw, {:ecto_rollback, value} ->
-        do_rollback(repo, worker, opts)
+        do_rollback(repo, key, Process.get(key), opts, timeout)
         {:error, value}
       type, term ->
         stacktrace = System.stacktrace
-        do_rollback(repo, worker, opts)
+        do_rollback(repo, key, Process.get(key), opts, timeout)
         :erlang.raise(type, term, stacktrace)
-    after
-      checkin_worker(pool, timeout)
+    else
+      value ->
+        do_commit(repo, key, Process.get(key), opts, timeout)
+        {:ok, value}
     end
   end
 
-  defp checkout_worker(pool, timeout) do
-    key = {:ecto_transaction_pid, pool}
-
-    case Process.get(key) do
-      {worker, counter} ->
-        Process.put(key, {worker, counter + 1})
-        worker
-      nil ->
-        worker = :poolboy.checkout(pool, true, timeout)
-        Worker.link_me(worker, timeout)
-        Process.put(key, {worker, 1})
-        worker
-    end
+  defp do_begin(repo, key, %{transactions: [], depth: 0, module: module} = info,
+  opts, timeout) do
+    info = %{info | transactions: [next: :transaction], depth: 1}
+    _ = Process.put(key, info)
+    query = apply(module, :begin_transaction, [])
+    log(repo, {:query, query, []}, opts, fn ->
+      begin_query!(key, info, query, [], opts, timeout)
+    end)
+    :ok
   end
-
-  defp checkin_worker(pool, timeout) do
-    key = {:ecto_transaction_pid, pool}
-
-    case Process.get(key) do
-      {worker, 1} ->
-        Worker.unlink_me(worker, timeout)
-        :poolboy.checkin(pool, worker)
-        Process.delete(key)
-      {worker, counter} ->
-        Process.put(key, {worker, counter - 1})
-    end
-
+  defp do_begin(repo, key, %{transactions: [_|_] = trans, depth: n, module: module} = info,
+  opts, timeout) do
+    savepoint = "ecto_#{n}"
+    n = n + 1
+    query = apply(module, :savepoint, [savepoint])
+    info = %{info | transactions: [next: {:savepoint, savepoint}] ++ trans, depth: n}
```

Comment:
Why do we need to keep the `trans` history? Wouldn't checking the depth be enough?